libsplinter/src/auth/oauth/builder/openid.rs

```diff
@@ -0,0 +1,136 @@
+// Copyright 2018-2020 Cargill Incorporated
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use reqwest::blocking::Client;
+
+use crate::auth::oauth::{builder::OAuthClientBuilder, error::OAuthClientBuildError, OAuthClient};
+use crate::auth::rest_api::identity::{openid::OpenIdUserIdentityProvider, IdentityProvider};
+use crate::error::{InternalError, InvalidStateError};
+
+/// Builds a new `OAuthClient` using an OpenID discovery document.
+///
+/// # Arguments
+///
+/// * `client_id` - The OpenID provider OAuth client ID
+/// * `client_secret` - The OpenID provider OAuth client secret
+/// * `redirect_url` - The endpoint that the OpenID provider will redirect to after it has completed
+///   authorization
+/// * `discovery_url` - The OpenID provider's discovery document
+pub struct OpenIdOAuthClientBuilder {
+    openid_discovery_url: Option<String>,
+    inner: OAuthClientBuilder,
+}
+
+impl OpenIdOAuthClientBuilder {
+    pub fn new() -> Self {
+        Self {
+            openid_discovery_url: None,
+            inner: OAuthClientBuilder {
+                ..Default::default()
+            },
+        }
+    }
+
+    /// Sets the client ID for the OAuth2 provider.
+    pub fn with_client_id(self, client_id: String) -> Self {
+        Self {
+            openid_discovery_url: self.openid_discovery_url,
+            inner: self.inner.with_client_id(client_id),
+        }
+    }
+
+    /// Sets the client secret for the OAuth2 provider.
+    pub fn with_client_secret(self, client_secret: String) -> Self {
+        Self {
+            openid_discovery_url: self.openid_discovery_url,
+            inner: self.inner.with_client_secret(client_secret),
+        }
+    }
+
+    /// Sets the redirect URL for the OAuth2 provider.
+    pub fn with_redirect_url(self, redirect_url: String) -> Self {
+        Self {
+            openid_discovery_url: self.openid_discovery_url,
+            inner: self.inner.with_redirect_url(redirect_url),
+        }
+    }
+
+    pub fn with_discovery_url(mut self, discovery_url: String) -> Self {
+        self.openid_discovery_url = Some(discovery_url);
+
+        self
+    }
+
+    /// Builds an OAuthClient and an [`IdentityProvdier`] based on the OpenID provider's discovery
+    /// document.
+    ///
+    /// # Errors
+    ///
+    /// Returns an [`OAuthClientBuildError`] if there are required fields missing, if any URL's
+    /// provided are invalid,  or if it is unable to load the discovery document..
```

Comment:
Double space on this line, also double . at the end.

Reply:
Fixed